book/18-concurrent-programming.md

```diff
@@ -0,0 +1,1062 @@
+# Concurrent Programming with Async {#concurrent-programming-with-async data-type="chapter"}
+
+ The logic of building programs that interact with the outside world is often
+dominated by waiting; waiting for the click of a mouse, or for data to be
+fetched from disk, or for space to be available on an outgoing network
+buffer. Even mildly sophisticated interactive applications are typically
+*concurrent*, needing to wait for multiple different events at the same time,
+responding immediately to whatever event happens first. <idx>interactive
+input/concurrent programming for</idx><idx>concurrent
+programming</idx><idx>programming/concurrent programming with Async</idx>
+
+ One approach to concurrency is to use preemptive system threads, which is
+the dominant approach in languages like Java or C#. In this model, each task
+that may require simultaneous waiting is given an operating system thread of
+its own so it can block without stopping the entire program.
+<idx>threads/preemptive vs. single-threaded programs</idx>
+
+ Another approach is to have a single-threaded program, where that single
+thread runs an *event loop* whose job is to react to external events like
+timeouts or mouse clicks by invoking a callback function that has been
+registered for that purpose. This approach shows up in languages like
+JavaScript that have single-threaded runtimes, as well as in many GUI
+toolkits. <idx>event loops</idx><idx>system threads</idx>
```

Comment:
It might be nice to have nicer way of expressing index terms (but it's not a big deal.)

Reply:
The purpose of these is to generate an index. So they shouldn't typically be rendered in the final text directly, but instead are fed to an index-generation process. We did this as part of generating the original O'Reilly book (actually, they wrote the index for us), and it would be nice not to lose that work, and indeed, to be able to help in updating and generating the index as we work on the book.

[event loops]{.idx} seems like an OK syntax for these.